data-document-model/src/main/java/io/micronaut/data/document/model/query/builder/CosmosSqlQueryBuilder2.java

```diff
@@ -0,0 +1,398 @@
+/*
+ * Copyright 2017-2022 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.data.document.model.query.builder;
+
+import io.micronaut.core.annotation.AnnotationMetadata;
+import io.micronaut.core.annotation.Creator;
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.annotation.NonNull;
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.core.util.ArgumentUtils;
+import io.micronaut.core.util.CollectionUtils;
+import io.micronaut.data.annotation.repeatable.WhereSpecifications;
+import io.micronaut.data.model.Association;
+import io.micronaut.data.model.Embedded;
+import io.micronaut.data.model.Pageable;
+import io.micronaut.data.model.Pageable.Mode;
+import io.micronaut.data.model.PersistentAssociationPath;
+import io.micronaut.data.model.PersistentEntity;
+import io.micronaut.data.model.PersistentPropertyPath;
+import io.micronaut.data.model.jpa.criteria.impl.selection.AggregateExpression;
+import io.micronaut.data.model.naming.NamingStrategies;
+import io.micronaut.data.model.naming.NamingStrategy;
+import io.micronaut.data.model.query.JoinPath;
+import io.micronaut.data.model.query.builder.QueryParameterBinding;
+import io.micronaut.data.model.query.builder.QueryResult;
+import io.micronaut.data.model.query.builder.sql.SqlQueryBuilder2;
+import jakarta.persistence.criteria.Expression;
+import jakarta.persistence.criteria.Predicate;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.StringJoiner;
+
+/**
+ * The Azure Cosmos DB sql query builder.
+ *
+ * @author radovanradic
+ * @since 3.9.0
+ */
+@Internal
+public final class CosmosSqlQueryBuilder2 extends SqlQueryBuilder2 {
```

Comment:
I would rename this hierarchy `LegacySqlQueryBuilder`, adding the number `2` is confusing IMO

Reply:
In that case it would be `CosmosSqlQueryBuilderNext`

Reply:
But that's only a temporary name till v5